src/gui/editorwidgets/qgsactionwidgetwrapper.cpp

```diff
@@ -0,0 +1,112 @@
+  qgsactionwidgetwrapper.cpp - QgsActionWidgetWrapper
+
+ ---------------------
+ begin                : 14.8.2021
+ copyright            : (C) 2021 by Alessandro Pasotti
+ email                : elpaso at itopen dot it
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#include "qgsactionwidgetwrapper.h"
+#include "qgsactionmanager.h"
+#include "qgsexpressioncontextutils.h"
+
+#include <QLayout>
+
+QgsActionWidgetWrapper::QgsActionWidgetWrapper( QgsVectorLayer *layer, QWidget *editor, QWidget *parent )
+  : QgsWidgetWrapper( layer, editor, parent )
+{
+  connect( this, &QgsWidgetWrapper::contextChanged, [ = ]
+  {
+    const bool actionIsVisible {
+      ( context().attributeFormMode() == QgsAttributeEditorContext::Mode::SingleEditMode ) ||
+      ( context().attributeFormMode() == QgsAttributeEditorContext::Mode::AddFeatureMode ) };
+    if ( mActionButton )
+    {
+      mActionButton->setVisible( actionIsVisible );
+    }
+  } );
+}
+
+void QgsActionWidgetWrapper::setAction( const QgsAction &action )
+{
+  mAction = action;
+}
+
+void QgsActionWidgetWrapper::setFeature( const QgsFeature &feature )
+{
+  mFeature = feature;
+}
+
+void QgsActionWidgetWrapper::setEnabled( bool enabled )
+{
+  if ( valid() && layer() )
+  {
+    mActionButton->setEnabled( ! mAction.isEnabledOnlyWhenEditable() || enabled );
+  }
+}
+
+bool QgsActionWidgetWrapper::valid() const
+{
+  return mAction.isValid() && mAction.runable();
+}
+
+QWidget *QgsActionWidgetWrapper::createWidget( QWidget *parent )
+{
+  return new QPushButton( parent );
+}
+
+void QgsActionWidgetWrapper::initWidget( QWidget *editor )
+{
+
+  mActionButton = qobject_cast<QPushButton *>( editor );
+
+  if ( !mActionButton )
+    return;
+
+  if ( valid() && layer() )
+  {
+    const QString shortTitle { mAction.shortTitle() }; // might be empty
+    const QString description { mAction.name() };  // mandatory
+    const QIcon icon { mAction.icon() };  // might be invalid
+
+    // Configure push button
+    if ( ! icon.isNull() )
+    {
+      mActionButton->setIcon( mAction.icon() );
+      mActionButton->setToolTip( description );
+    }
+    else
+    {
+      mActionButton->setText( shortTitle.isEmpty() ? description : shortTitle );
+      if ( ! shortTitle.isEmpty() )
+      {
+        mActionButton->setToolTip( description );
+      }
+    }
+
+    if ( mAction.isEnabledOnlyWhenEditable() && ! layer()->isEditable() )
```

Comment:
```suggestion
    if ( mAction.isEnabledOnlyWhenEditable() && !layer()->isEditable() )
```

Reply:
also, following the logic further down, you should also check for the form mode?
(see https://github.com/qgis/QGIS/pull/44710/files#diff-63b2666af07d22ef212357c4b9c789a154f4c54a35429d09f0800471c1304bafR1345)

Reply:
I originally did that, but I changed it when I realized (if I'm not wrong) that the mode can change without `initWidget` being called again, the lambda in the ctor takes care of updating the widget status (visibility) dynamically whenever the form mode changes. The editable state is kept in sync by `QgsActionWidgetWrapper::setEnabled`